flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java

```diff
@@ -0,0 +1,147 @@
+    }
+
+    /**
+     * Mark splits with specified ids as finished.
+     *
+     * @param finishedSplitIds collection of split ids to mark as finished
+     */
+    public void markAsFinished(Collection<String> finishedSplitIds) {
+        finishedSplitIds.forEach(
+                splitId -> {
+                    assignedSplits.remove(splitId);
+                    knownSplits.remove(splitId);
+                });
+    }
+
+    /**
+     * Checks if split with specified id had been assigned to the reader.
+     *
+     * @param splitId split id
+     * @return {@code true} if split had been assigned, otherwise {@code false}
+     */
+    public boolean isAssigned(String splitId) {
+        return assignedSplits.contains(splitId);
+    }
+
+    /**
+     * Returns list of splits that can be assigned to readers. Does not include splits that are
+     * already assigned or finished. If shard ordering is enabled, only splits with finished parents
+     * will be returned.
+     *
+     * @return list of splits that can be assigned to readers.
+     */
+    public List<KinesisShardSplit> splitsAvailableForAssignment() {
+        return knownSplits.values().stream()
+                .filter(
+                        split -> {
+                            boolean splitIsNotAssigned = !isAssigned(split.splitId());
+                            if (preserveShardOrdering) {
+                                // Check if all parent splits were finished
+                                return splitIsNotAssigned
+                                        && verifyAllParentSplitsAreFinished(split);
+                            } else {
+                                return splitIsNotAssigned;
+                            }
+                        })
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Prepare split graph representation to store in state. Method returns only splits that are
+     * currently assigned to readers or unassigned. Finished splits are not included in the result.
+     *
+     * @param checkpointId id of the checkpoint
+     * @return list of splits with current assignment status
+     */
+    public List<KinesisShardSplitWithAssignmentStatus> snapshotState(long checkpointId) {
+        return knownSplits.values().stream()
+                .map(
+                        split -> {
+                            SplitAssignmentStatus assignmentStatus =
+                                    isAssigned(split.splitId())
+                                            ? SplitAssignmentStatus.ASSIGNED
+                                            : SplitAssignmentStatus.UNASSIGNED;
+                            return new KinesisShardSplitWithAssignmentStatus(
+                                    split, assignmentStatus);
+                        })
+                .collect(Collectors.toList());
+    }
+
+    private boolean verifyAllParentSplitsAreFinished(KinesisShardSplit split) {
```

Comment:
Noted that this implementation also covers the edge case, when we restore from a stale snapshot that is taken when not consuming at the tip of stream. Let's make sure we have a test case for this!

### Edge case 1
The edge case I'm thinking of:
```
A ---\   /--- E
       C 
B ---/   \--- F
```
Above I show a shard that was merged, then split. Although this is unrealistic in a 2 shard system, it represents a situation that can easily occur in large shard systems.

First we imagine a snapshot is taken when stream is aware of A,B,C, but currently consuming from A+B.
```
Current stream condition:
A ---\   
      C 
B ---/   

Enumerator state:
knownSplits: A,B,C
assignedSplits: A,B
```

Ok imagine now we keep this snapshot until the stream has expired shards A + B + C

```
Current stream condition:
  /--- E
C (expired)
  \--- F
```

Now, what will happen?
1. We discover splits E + F.
```
Enumerator state:
knownSplits: C,E (waiting on C), F (waiting on C)
assignedSplits: A,B
```
2. Assigned splits A + B will be marked as finished because they are expired. Assumed behavior of the SplitReader.

```
Enumerator state:
knownSplits: C,E (waiting on C), F (waiting on C)
assignedSplits: 
```
3. Enumerator will assign split C.
4. Split C will be marked as finished because they are expired. (Once again assumed behavior of SplitReader)
5. We will then start consuming from E + F.



### Edge case 2

```
A ---\   /--- E ---\
       C            G
B ---/   \--- F ---/
```


First we imagine a snapshot is taken when stream is aware of A,B,C, but currently consuming from A+B.
```
Current stream condition:
A ---\   
      C 
B ---/   

Enumerator state:
knownSplits: A,B,C
assignedSplits: A,B
```

Then we keep snapshot until it is stale, and shards A,B,C,E,F have expired.

```
Current stream condition:
E (expired) ---\
                G
F (expired) ---/

Enumerator state:
knownSplits: A,B,C
assignedSplits: A,B
```

Now, what will happen?
1. We discover splits G.
```
Enumerator state:
knownSplits: A,B,C,G(waiting on E,F)
assignedSplits: A,B
```
2. Assigned splits A + B will be marked as finished because they are expired. Assumed behavior of the SplitReader.

```
Enumerator state:
knownSplits: C,G(waiting on E,F)
assignedSplits: 
```
3. Enumerator will assign split C + G
4. Split C will be marked as finished because they are expired. (Once again assumed behavior of SplitReader). We will start consuming from G directly.